kotlinx-coroutines-core/jvm/test/ThreadLocalStressTest.kt

```diff
@@ -63,10 +67,99 @@ class ThreadLocalStressTest : TestBase() {
             withContext(threadLocal.asContextElement("foo")) {
                 yield()
                 cancel()
-                suspendCancellableCoroutineReusable<Unit> {  }
+                suspendCancellableCoroutineReusable<Unit> { }
             }
         } finally {
             assertEquals(expectedValue, threadLocal.get())
         }
     }
+
+    /*
+     * Another set of tests for undispatcheable continuations that do not require stress test multiplier.
+     * Also note that `uncaughtExceptionHandler` is used as the only available mechanism to propagate error from
+     * `resumeWith`
+     */
+
+    @Test
+    fun testNonDispatcheableLeak() {
+        repeat(100) {
+            doTestWithPreparation(
+                ::doTest,
+                { threadLocal.set(null) }) { threadLocal.get() != null }
+            assertNull(threadLocal.get())
+        }
+    }
+
+    @Test
+    fun testNonDispatcheableLeakWithInitial() {
+        repeat(100) {
+            doTestWithPreparation(::doTest, { threadLocal.set("initial") }) { threadLocal.get() != "initial" }
+            assertEquals("initial", threadLocal.get())
+        }
+    }
+
+    @Test
+    fun testNonDispatcheableLeakWithContextSwitch() {
+        repeat(100) {
+            doTestWithPreparation(
+                ::doTestWithContextSwitch,
+                { threadLocal.set(null) }) { threadLocal.get() != null }
+            assertNull(threadLocal.get())
+        }
+    }
+
+    @Test
+    fun testNonDispatcheableLeakWithInitialWithContextSwitch() {
+        repeat(100) {
+            doTestWithPreparation(
+                ::doTestWithContextSwitch,
+                { threadLocal.set("initial") }) { false /* can randomly wake up on the non-main thread */ }
+            // Here we are always on the main thread
+            assertEquals("initial", threadLocal.get())
+        }
+    }
+
+    private fun doTestWithPreparation(testBody: suspend () -> Unit, setup: () -> Unit, isInvalid: () -> Boolean) {
```

Comment:
Minor, opinionated: `isInvalid` is a reverse of `isValid`, and most implementations here are of the form `!=` or `false`, which is a double negation, adding mental load to this and making code asymmetrical and, arguably, even non-idiomatic, as asserts are most of the time state success conditions positive.